datafusion/src/datasource/listing/helpers.rs

```diff
@@ -0,0 +1,722 @@
+            // TODO: Stable functions could be `applicable`, but that would require access to the context
+            Volatility::Stable | Volatility::Volatile => {
+                *self.is_applicable = false;
+                Recursion::Stop(self)
+            }
+        }
+    }
+}
+
+impl ExpressionVisitor for ApplicabilityVisitor<'_> {
+    fn pre_visit(self, expr: &Expr) -> Result<Recursion<Self>> {
+        let rec = match expr {
+            Expr::Column(logical_plan::Column { ref name, .. }) => {
+                *self.is_applicable &= self.col_names.contains(name);
+                Recursion::Stop(self) // leaf node anyway
+            }
+            Expr::Literal(_)
+            | Expr::Alias(_, _)
+            | Expr::ScalarVariable(_)
+            | Expr::Not(_)
+            | Expr::IsNotNull(_)
+            | Expr::IsNull(_)
+            | Expr::Negative(_)
+            | Expr::Cast { .. }
+            | Expr::TryCast { .. }
+            | Expr::BinaryExpr { .. }
+            | Expr::Between { .. }
+            | Expr::InList { .. }
+            | Expr::GetIndexedField { .. }
+            | Expr::Case { .. } => Recursion::Continue(self),
+
+            Expr::ScalarFunction { fun, .. } => self.visit_volatility(fun.volatility()),
+            Expr::ScalarUDF { fun, .. } => {
+                self.visit_volatility(fun.signature.volatility)
+            }
+
+            // TODO other expressions are not handled yet:
+            // - AGGREGATE, WINDOW and SORT should not end up in filter conditions, except maybe in some edge cases
+            // - Can `Wildcard` be considered as a `Literal`?
+            // - ScalarVariable could be `applicable`, but that would require access to the context
+            Expr::AggregateUDF { .. }
+            | Expr::AggregateFunction { .. }
+            | Expr::Sort { .. }
+            | Expr::WindowFunction { .. }
+            | Expr::Wildcard => {
+                *self.is_applicable = false;
+                Recursion::Stop(self)
+            }
+        };
+        Ok(rec)
+    }
+}
+
+/// Check whether the given expression can be resolved using only the columns `col_names`.
+/// This means that if this function returns true:
+/// - the table provider can filter the table partition values with this expression
+/// - the expression can be marked as `TableProviderFilterPushDown::Exact` once this filtering
+/// was performed
+pub fn expr_applicable_for_cols(col_names: &[String], expr: &Expr) -> bool {
+    let mut is_applicable = true;
+    expr.accept(ApplicabilityVisitor {
+        col_names,
+        is_applicable: &mut is_applicable,
+    })
+    .unwrap();
+    is_applicable
+}
+
+/// Partition the list of files into `n` groups
+pub fn split_files(
+    partitioned_files: Vec<PartitionedFile>,
+    n: usize,
+) -> Vec<Vec<PartitionedFile>> {
+    if partitioned_files.is_empty() {
+        return vec![];
+    }
+    let chunk_size = (partitioned_files.len() + n - 1) / n;
```

Comment:
```suggestion
    // effectively this is div with rounding up instead of truncating
    let chunk_size = (partitioned_files.len() + n - 1) / n;
```